haystack/document_stores/opensearch.py

```diff
@@ -699,24 +872,62 @@ def _get_embedding_field_mapping(
+
+            ef_construction = (
+                80 if "ef_construction" not in self.knn_parameters else self.knn_parameters["ef_construction"]
+            )
+            ef_search = self._get_ef_search_value()
+            m = 64 if "m" not in self.knn_parameters else self.knn_parameters["m"]
 
             if index_type == "flat":
+                # We're using HNSW with knn_engines nmslib and faiss as they do not support exact knn.
+                method["name"] = "hnsw"
                 # use default parameters from https://opensearch.org/docs/1.2/search-plugins/knn/knn-index/
                 # we need to set them explicitly as aws managed instances starting from version 1.2 do not support empty parameters
                 method["parameters"] = {"ef_construction": 512, "m": 16}
             elif index_type == "hnsw":
-                method["parameters"] = {"ef_construction": 80, "m": 64}
+                method["name"] = "hnsw"
+                method["parameters"] = {"ef_construction": ef_construction, "m": m}
                 # for nmslib this is a global index setting
                 if knn_engine == "faiss":
-                    method["parameters"]["ef_search"] = 20
+                    method["parameters"]["ef_search"] = ef_search
+            elif index_type in ["ivf", "ivf_pq"]:
+                if knn_engine != "faiss":
+                    raise DocumentStoreError("To use 'ivf' or 'ivf_pq as index_type, set knn_engine to 'faiss'.")
+                # Check if IVF model already exists
+                if self._index_exists(".opensearch-knn-models"):
+                    response = self.client.transport.perform_request("GET", "/_plugins/_knn/models/_search")
+                    existing_ivf_models = set(
+                        model["_source"]["model_id"]
+                        for model in response["hits"]["hits"]
+                        if model["_source"]["state"] == "created"
+                        and model["_source"]["dimension"] == embedding_dim
+                        and model["_source"]["space_type"] == space_type
+                    )
+                else:
+                    existing_ivf_models = set()
+                if f"{index}-ivf" in existing_ivf_models:
+                    logger.info("Using existing IVF model '%s-ivf' for index '%s'.", index, index)
+                    embeddings_field_mapping = {"type": "knn_vector", "model_id": f"{index}-ivf"}
+                    method = {}
+                else:
+                    # IVF indices require training before they can be initialized. Setting index_type to HNSW until
+                    # index is trained
+                    logger.info("Using index of type HNSW for index '%s' until IVF model is trained.", index)
+                    method["name"] = "hnsw"
+                    method["parameters"] = {"ef_construction": ef_construction, "m": m, "ef_search": ef_search}
```

Comment:
I wonder whether we should use the "hnsw" or the "flat" settings here. If I get it correctly, it's not being used anyways. If I try to add documents via `write_documents()` the IVF would get created. "flat" values would be the default values of OpenSearch which is somewhat more understandable here.

Reply:
I decided to go for an exact index using score_script instead of HNSW until the IVF index is trained. No need to construct an approximate HNSW index if we’re not not allowing to query it.